x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/repositories/GetCcrRestoreFileChunkAction.java

```diff
@@ -79,24 +77,20 @@ public void onFailure(Exception e) {
                 @Override
                 protected void doRun() throws Exception {
                     int bytesRequested = request.getSize();
-                    long fileOffset = request.getOffset();
                     ByteArray array = bigArrays.newByteArray(bytesRequested, false);
                     String fileName = request.getFileName();
                     String sessionUUID = request.getSessionUUID();
                     // This is currently safe to do because calling `onResponse` will serialize the bytes to the network layer data
                     // structure on the same thread. So the bytes will be copied before the reference is released.
                     try (ReleasablePagedBytesReference reference = new ReleasablePagedBytesReference(array, bytesRequested, array)) {
                         try (CcrRestoreSourceService.Reader reader = restoreSourceService.getSessionReader(sessionUUID, fileName)) {
-                            BytesRefIterator refIterator = reference.iterator();
-                            BytesRef ref;
-                            int bytesWritten = 0;
-                            while ((ref = refIterator.next()) != null) {
-                                byte[] refBytes = ref.bytes;
-                                reader.readFileBytes(refBytes, fileOffset + bytesWritten, ref.length);
-                                bytesWritten += ref.length;
+                            int bytesRead = reader.readFileBytes(reference);
+                            if (bytesRead < bytesRequested) {
```

Comment:
Also do we need to protect this from being called concurrently?

Reply:
We now synchronize all read file bytes.

```
doesn't this mean we need to exactly know how much is left to read otherwise bytesRequested will be bigger than the remaining bytes in the case we are approaching the end of the file? Can't we simply treat this as an upperbound? seems odd to fail here.
```

We do know how much we need to read. The follower knows the size of the file as it requests chunks. This was mainly a sanity check to ensure we did not read less than expected mainly due to iteration over byterefs (like what if we read into half a ref, then read into the next ref. There would be empty bytes in the middle of the reference). However, I see that `DataOutput` will throw a `EOFException` if you attempt to read too much. So this sanity check can be removed.